src/ifc5d/ifc5d/qto.py

```diff
@@ -78,41 +78,52 @@ def visit(decl):
+    # Precompute entity names once
+    schema_id = ifc_file.schema_identifier
+    entity_names_lower = lower_case_entity_names(schema_id)
 
-    :param rules: Set of rules from `ifc5d.qto.rules`.
-    """
-    results: ResultsDict = {}
-    elements_by_classes: defaultdict[str, set[ifcopenshell.entity_instance]] = defaultdict(set)
+    # Pre-group elements by type (avoids sorting/groupby)
+    elements_by_class: dict[str, list[ifcopenshell.entity_instance]] = defaultdict(list)
     for element in elements:
-        elements_by_classes[element.is_a()].add(element)
-
-    for calculator, queries in rules["calculators"].items():
-        calculator = calculators[calculator]
-        if not set(m.lower() for m in queries.keys()) - lower_case_entity_names(ifc_file.schema_identifier):
-            # all defined queries are actually simple entity names: instead of looping over all
-            # calculators we loop over the entity types so that we don't have to repeatedly query
-            # the model, especially when the set of elements is small.
+        elements_by_class[element.is_a().lower()].append(element)
+
+    for calculator_name, queries in rules["calculators"].items():
+        calculator = calculators[calculator_name]
+        queries_lower = {k.lower(): v for k, v in queries.items()}
+
+        # Determine if all queries are direct entity types (simple case)
+        if not (set(queries_lower.keys()) - entity_names_lower):
             casenorm = {k.lower(): k for k in queries.keys()}
-            pred = lambda inst: inst.is_a().lower()
-            for ty, elements in itertools.groupby(sorted(elements, key=pred), key=pred):
-                for sty in entity_supertypes(ifc_file.schema_identifier, ty):
-                    if qtos := queries.get(casenorm.get(sty)):
-                        calculator.calculate(ifc_file, list(elements), qtos, results)
-        for query, qtos in queries.items():
-            if query.lower() in lower_case_entity_names(ifc_file.schema_identifier):
-                # by_type is faster than selector parsing
-                filtered_elements = ifc_file.by_type(query)
-            else:
-                filtered_elements = ifcopenshell.util.selector.filter_elements(ifc_file, query, elements)
-            if filtered_elements:
-                calculator.calculate(ifc_file, filtered_elements, qtos, results)
+            # Cache supertypes per entity to avoid recomputation
+            supertype_cache = {}
+
+            for entity_type, elems in elements_by_class.items():
+                if entity_type not in supertype_cache:
+                    supertype_cache[entity_type] = entity_supertypes(schema_id, entity_type)
+
+                for sty in supertype_cache[entity_type]:
+                    if (sty in casenorm) and (qtos := queries.get(casenorm[sty])):
+                        calculator.calculate(ifc_file, elems, qtos, results)
+        else:
```

Comment:
I think the important thing is the else statement here. That is really crucial otherwise we're doing it twice in the simple case.